src/util/type.cpp

```diff
@@ -34,3 +35,72 @@ bool is_number(const typet &type)
          id==ID_floatbv ||
          id==ID_fixedbv;
 }
+
+/// Identify if a given type is constant itself or
+/// contains constant components. Examples include:
+///  - const int a;
+///  - struct contains_constant_pointer {  int x; int * const p; };
+///  - const int b[3];
+/// \param type The type we want to query constness of.
+/// \param ns The namespace, needed for resolution of symbols.
+/// \return Whether passed in type is const or not.
+bool is_constant_or_has_constant_components(
+  const typet &type, const namespacet &ns)
+{
+  // Helper function to avoid the code duplication in the branches
+  // below.
+  const auto has_constant_components =
+    [&ns](const typet &subtype) -> bool
+    {
+      if(subtype.id() == ID_struct || subtype.id() == ID_union)
+      {
+        const auto &struct_union_type = to_struct_union_type(subtype);
+        for(const auto &component : struct_union_type.components())
+        {
+          if(is_constant_or_has_constant_components(component.type(), ns))
+            return true;
+        }
+      }
+      return false;
+    };
+
+  // There are 4 possibilities the code below is handling.
+  // The possibilities are enumerated as comments, to show
+  // what each code is supposed to be handling. For more
+  // comprehensive test case for this, take a look at
+  // regression/cbmc/no_nondet_static/main.c
+
+  // const int a;
+  if(type.get_bool(ID_C_constant))
+    return true;
+
+  // Handles the case where we have a recursive data type definition
+  // like the linked list below.
+  // struct list { const int datum; struct list * next; };
```

Comment:
This comment isn't strictly true - this case doesn't handle recursive structs, but I think what you meant is that this case is used to terminate the recursion in cases where the struct is a recursive structure. Might also be worth highlighting the fairly subtle distinction between this case (where for pointers you *always* return) and the case in lines 74/75 - where you only return if the type is a constant (and fall through to this case if not).